compiler/noirc_evaluator/src/ssa/ir/instruction.rs

```diff
@@ -818,25 +818,87 @@ fn try_optimize_array_get_from_previous_set(
+/// to `v1` and see that the previous array set occurs to a different constant index.
+///
+/// For each array set:
+/// - If the index is non-constant we fail the optimization since any index may be changed.
+/// - If the index is constant and is our target index, we conservatively fail the optimization.
+/// - Otherwise, we check the array value of the array set. We will refer to this array as array'.
+///   In the case above, array' is `v1` from `v5 = array set ...`
+///   - Check the original array set value's array get against array'. If they are equal we can simplify.
+///     - Continuing the example above, as we have `v3 = array_get v1, index 1`, `v1` is
+///       what we want to check against array'. We now know we can simplify `v7` to `v5` as it is unchanged.
+///   - If they are not equal, recur
 fn try_optimize_array_set_from_previous_get(
     dfg: &DataFlowGraph,
-    array_id: ValueId,
+    mut array_id: ValueId,
     target_index: ValueId,
     target_value: ValueId,
 ) -> SimplifyResult {
-    match &dfg[target_value] {
+    let array_from_get = match &dfg[target_value] {
         Value::Instruction { instruction, .. } => match &dfg[*instruction] {
             Instruction::ArrayGet { array, index } => {
                 if *array == array_id && *index == target_index {
-                    SimplifyResult::SimplifiedTo(array_id)
+                    // If array and index match from the value, we can immediately simplify
+                    return SimplifyResult::SimplifiedTo(array_id);
+                } else if *index == target_index {
+                    *array
                 } else {
-                    SimplifyResult::None
+                    return SimplifyResult::None;
                 }
             }
-            _ => SimplifyResult::None,
+            _ => return SimplifyResult::None,
         },
-        _ => SimplifyResult::None,
+        _ => return SimplifyResult::None,
+    };
+
+    let Some(target_index) = dfg.get_numeric_constant(target_index) else {
+        return SimplifyResult::None;
+    };
+
+    let original_array_id = array_id;
+    // Arbitrary number of maximum tries just to prevent this optimization from taking too long.
+    let max_tries = 5;
+    for _ in 0..max_tries {
+        match &dfg[array_id] {
+            Value::Instruction { instruction, .. } => match &dfg[*instruction] {
+                Instruction::ArraySet { array, index, .. } => {
+                    if let Some(index) = dfg.get_numeric_constant(*index) {
+                        if index == target_index {
+                            return SimplifyResult::None;
+                        }
+
+                        if *array == array_from_get {
+                            return SimplifyResult::SimplifiedTo(original_array_id);
+                        }
+
+                        array_id = *array; // recur
+                    } else {
+                        return SimplifyResult::None;
+                    }
```

Comment:
```suggestion
                    let Some(index) = dfg.get_numeric_constant(*index) else  {
                        return SimplifyResult::None;
                    }    
                    
                    if index == target_index {
                        return SimplifyResult::None;
                    }

                    if *array == array_from_get {
                        return SimplifyResult::SimplifiedTo(original_array_id);
                    }

                    array_id = *array; // recur
```

We could also remove a couple of layers from the match statements in a similar fashion tbh.